std/algorithm/comparison.d

```diff
@@ -644,66 +672,33 @@ if (isInputRange!R1 && isInputRange!R2)
     }
     else
     {
-        import core.stdc.string : memcmp;
-        import std.utf : decode;
-
-        // For speed only
-        static int threeWay(size_t a, size_t b)
-        {
-            static if (size_t.sizeof == int.sizeof)
-                return a - b;
-            else
-                // Faster than return b < a ? 1 : a < b ? -1 : 0;
-                return (a > b) - (a < b);
-        }
-        // For speed only
-        // @@@BUG@@@ overloading should be allowed for nested functions
-        static int threeWayInt(int a, int b)
-        {
-            return a - b;
-        }
-
         static if (typeof(r1[0]).sizeof == typeof(r2[0]).sizeof)
         {
-            static if (typeof(r1[0]).sizeof == 1)
-            {
-                immutable len = min(r1.length, r2.length);
-                int result = __ctfe ?
-                    {
-                        foreach (i; 0 .. len)
-                        {
-                            if (r1[i] != r2[i])
-                                return threeWayInt(r1[i], r2[i]);
-                        }
-                        return 0;
-                    }()
-                    : () @trusted { return memcmp(r1.ptr, r2.ptr, len); }();
-                if (result) return result;
-                return threeWay(r1.length, r2.length);
-            }
-            else
+            return () @trusted
             {
-                return () @trusted
+                auto p1 = r1.ptr, p2 = r2.ptr,
+                    pEnd = p1 + min(r1.length, r2.length);
+                for (; p1 != pEnd; ++p1, ++p2)
                 {
-                    auto p1 = r1.ptr, p2 = r2.ptr,
-                        pEnd = p1 + min(r1.length, r2.length);
-                    for (; p1 != pEnd; ++p1, ++p2)
-                    {
-                        if (*p1 != *p2) return threeWayInt(int(*p1), int(*p2));
-                    }
-                    return threeWay(r1.length, r2.length);
-                }();
-            }
+                    if (*p1 != *p2) return cast(int) *p1 - cast(int) *p2;
```

Comment:
you should cache `*p1` and `*p2` 
the additional dereference may be expensive

Reply:
@stefan-koch-sociomantic  I've verified that the compiler performs this optimization without needing any hint from the programmer. Both `cmp1` and `cmp1` in the example below are compiled to the same machine code by `ldc -O2`:

https://run.dlang.io/is/cz5hay
```d
import std.algorithm : min;

int cmp1(const dchar[] r1, const dchar[] r2) @trusted
{
    auto p1 = r1.ptr, p2 = r2.ptr, pEnd = p1 + min(r1.length, r2.length);
    for (; p1 != pEnd; ++p1, ++p2)
    {
        if (*p1 != *p2)
            return cast(int)*p1 - cast(int)*p2;
    }
    return int(r1.length > r2.length) - int(r1.length < r2.length);
}

int cmp2(const dchar[] r1, const dchar[] r2) @trusted
{
    auto p1 = r1.ptr, p2 = r2.ptr, pEnd = p1 + min(r1.length, r2.length);
    for (; p1 != pEnd; ++p1, ++p2)
    {
        const v1 = *p1, v2 = *p2; // manually saving result of *p1 and *p2
        if (v1 != v2)
            return cast(int) v1 - cast(int) v2;
    }
    return int(r1.length > r2.length) - int(r1.length < r2.length);
}

void main()
{
}
```